### PR TITLE
CoinValidator Refactoring

### PR DESCRIPTION
**Pull request by BelghmiAmine**

### CODE REVIEW

**Review by K1li4nL — DISMISSED** [review]
All good to me

**Comment by Ajkunas** [comment]
You need to change back your base branch to `work-be2-amine-ElectionHandlerRefactoring` because there is still the changes of the refactoring of MeetingValidator that appear.

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala
Comment (Ajkunas) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala:37]: Maybe some comments for these two new functions. :)

**Review by Ajkunas — CHANGES_REQUESTED** [review]
See my comments, there are some little changes to be done.

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala
Comment (Ajkunas) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala:57]: I think you can also give as a parameter the `error: PipelinError` to this function. Thus, the functions are consistant between themself.

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala
Comment (Ajkunas) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala:22]: Also try to split up those functions for readability.

**Review by Ajkunas — CHANGES_REQUESTED** [review]
Last little comments. I am happy to approve your PR after you take them in consideration. :)

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala
Comment (Ajkunas) [inline be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala:58]: You do not need to define this function inside. It is already defined in the class (see line 15). We simply redefine the validationError which concerns the message, thus we do it inside the class, because it is used in every validators.

Comment (Ajkunas) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala:58]: Mmh after those changes, I believe that the optimal implementation was to keep the same structure, aka: 
`transaction match {
            case Left(err) => Right(validationError(err.getMessage()))
            case Right(_)  => Left(rpcMessage)`

Sorry, I was not realizing that then you would need to redefine a variable :(
I think keeping it as a match case is also more clear. And then I guess you also do not need to put the `error: PipelineError` in the parameters (Sorry again :( )

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala
Comment (Ajkunas) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala:22]: I reiterate with this comment.

**Review by Ajkunas — CHANGES_REQUESTED** [review]
Okeey I made a bad suggestion previously, my bad ! 
I will let you see my comments, but the initial implementation of `checkSumOutputs` was way more optimal, just realizing it now. 

Do not forget to split up the functions with too many lines. 

After this, I am happy to approve your PR :)

File: be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala
Comment (BelghmiAmine) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala:58]: The code doesn't compile when I delete line 58 (def validationError()...
)

Comment (Ajkunas) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala:54]: Mmmh okeey I see now why it does not compile.. 
Indeed you need to access the `err` to use it inside the `validationError` function. 

For now, the parameter `error: PipelineError` is not used inside your function. Can't you try to give as a parameter the function `validationError`, and then use it inside. So, that you do not need to redefine a local one inside.

Comment (Ajkunas) [reply be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/CoinValidator.scala:54]: So, you delete the parameter `error: PipelineError` and give as a parameter `validationError: String => PipelineError` instead.

**Review by K1li4nL — APPROVED** [review]

**Pull request merged**
